lib/postcode_validation/use_case/validate_address.rb

```diff
@@ -11,10 +11,18 @@ def initialize(address_match_gateway:, logger: nil)
       def execute(postcode:, country:)
         @postcode = postcode
         @country = country
-        { valid?: valid_postcode? }
+        errors = []
+        errors << 'no_country_provided' if country.nil?
+        errors << 'invalid_format' if invalid_postcode_format?
+        matches = potential_address_matches
+        errors << 'no_matches' if matches.first.nil?
+        return { valid?: false, reason: errors } unless errors.empty?
```

Comment:
Wonder if we could push some of this stuff down into methods.  Can help you check it out if you want.